airflow/models/dag.py

```diff
@@ -1460,12 +1460,16 @@ def fetch_callback(
             dagrun = DAG.fetch_dagrun(dag_id=dag.dag_id, run_id=dag_run_id, session=session)
             callbacks = callbacks if isinstance(callbacks, list) else [callbacks]
             tis = dagrun.get_task_instances(session=session)
+            # Filter out removed tasks and
             # tis from a dagrun may not be a part of dag.partial_subset,
             # since dag.partial_subset is a subset of the dag.
             # This ensures that we will only use the accessible TI
             # context for the callback.
-            if dag.partial:
-                tis = [ti for ti in tis if not ti.state == State.NONE]
+            tis = [
+                ti
+                for ti in tis
+                if ti.state != TaskInstanceState.REMOVED or (dag.partial and not ti.state == State.NONE)
+            ]
```

Comment:
What you had earlier seems better, can we go back to it? Sorry for pushing you to this direction